File: 1122-relative-sort-array/1122-relative-sort-array.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        Map<Integer, Integer> cnt = new HashMap<>();
        int[] result = new int[arr1.length];
        PriorityQueue<Integer> left = new PriorityQueue<>();
        int idx = 0;

        for (int num: arr1) {
            cnt.put(num, cnt.getOrDefault(num, 0) + 1);
        }

        for (int num: arr2) {
            for (int i = 0; i < cnt.get(num); i++) {
                result[idx++] = num;
            }
            cnt.put(num, 0);
        }

        cnt.forEach((key, value) -> {
            for (int i = 0; i < value; i++) {
                left.add(key);
            }
        });

        while (!left.isEmpty()) {
            result[idx++] = left.poll();
        }

        return result;
    }
}

File: 1122-relative-sort-array/1122-relative-sort-array.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        cnt = defaultdict(int)
        result = list()
        left = list()

        for num in arr1:
            cnt[num] += 1

        for num in arr2:
            for i in range(cnt[num]):
                result.append(num)
            cnt[num] = 0

        for key, value in cnt.items():
            for i in range(value):
                heapq.heappush(left, key)

        while left:
            result.append(heapq.heappop(left))

        return result